Replace AssetTransferred with AllocationUpdated event

geoknee opened the pull request:
This emits a close-to-minimal amount of information when transfer or claim is called: the information is sufficient for clients to construct a new AssetOutcome for a given channel in a given AssetHolder. They need this up-to-date data in order to successfully call transfer in future.

Example 1: my counterparty concluded, pushed, and transferred their own money out of a channel. Now I want to transfer my money out, but my view of the outcome on chain is out of date and my transaction will fail if I don't keep it fresh.

Example 2: someone called `concludePushOutcomeAndTransferAll`, and I want to be sure all the money got paid out. It's not the only way of doing this, but I can catch the `AllocationUpdated` event, feed it into the new typescript functions this PR introduces, and infer which payouts have been made. 

- [x] contracts updated with new event type
- [x] off-chain version of _computeNewAllocation. We could go down the pureEVM route, but for now I just reimplemented it in typescript and
 wrote a test that uses random data to check against the solidity version.
- [x] existing tests updated to catch and assert on new event
- [x] AssetTransferred deprecated in nitro-protocol and the documentation
- [x] Chain Service updated to watch for the new event and call into computeNewAllocation to infer the data it wants to understand.
Closes #3046

geoknee commented on packages/server-wallet/src/wallet/wallet.ts:
```diff
@@ -660,16 +661,28 @@ export class SingleThreadedWallet extends EventEmitter<EventEmitterType>
     response.channelUpdatedEvents().forEach(event => this.emit('channelUpdated', event.value));
   }
 
-  async assetTransferred(arg: AssetTransferredArg): Promise<void> {
+  async assetOutcomeUpdated({
```
Something I spotted when doing this work, but is not really related to this PR: should this set of methods (which exist to implement the `ChainEventSubscriberInterface`) be private? It seems a little weird to me that they are a part of the Wallet API.

We could nest them all in a chainEventSubscriber property and pass that object into `this.chainService.registerChannel(...,...,)` instead of the entire wallet instance.

geoknee replied:
cc @kerzhner

andrewgordstewart reviewed (approved):
I think `computeNewAssetOutcome` could be made more understandable. Otherwise, this looks good.

I'm on board for random testing! Just make sure that in the event of a test failure, we get enough information to reproduce the test failure.

geoknee commented on packages/nitro-protocol/src/contract/asset-holder.ts:
```diff
@@ -49,3 +48,143 @@ export function convertAddressToBytes32(address: string): string {
+  assetHolderAddress: Address,
+  nitroAdjudicatorAddress: Address,
+  allocationUpdatedEvent: {channelId: Bytes32; initialHoldings: string},
+  tx?: ethers.Transaction
```
Yeah it shouldn't be optional, will fix.

geoknee commented:
@andrewgordstewart thanks for the detailed review. I think I addressed all of your points in 70c01d6 to 57f8357. I just need to solve the merge conflicts next.

geoknee commented on packages/server-wallet/src/chain-service/chain-service.ts:
```diff
@@ -297,6 +301,14 @@ export class ChainService implements ChainServiceInterface {
     return this.sendTransaction(pushTransactionRequest);
   }
 
+  // TODO add another public method for transferring from an channel that has already been concluded *and* pushed
+  // i.e. one that calls NitroAdjudicator.transfer. We could call the new method "withdraw" to match the other methods in this class
+  // The new method will uses the latest outcome that we know exists ON CHAIN, which can be "newer" than the latest OFF CHAIN outcome.
+  // Choices:
+  // - the chain service stores the latest outcome in it's own DB table
+  // - the wallet.store stores them
+  // - nothing is stored, and the chain service searches through historic transaction data that it pulls from the provider (seems riskier, but we may need to do this in case we are offline for a time, anyway)
+
```
https://github.com/statechannels/statechannels/issues/3076

geoknee commented:
Looking at the gas artifacts for `concludePushOutcomeAndTransferAll` on this[ branch](https://app.circleci.com/pipelines/github/statechannels/statechannels/11000/workflows/f54a542b-cb5b-49d7-ae83-ae7f34ba3ed5/jobs/53777/artifacts): - for 100 TOK payouts we need *2379417* gas . And on a recent, prior test run on [master ](https://app.circleci.com/pipelines/github/statechannels/statechannels/11000/workflows/f54a542b-cb5b-49d7-ae83-ae7f34ba3ed5/jobs/53777/artifacts) *2566892* gas.

This is consistent with saving approx 1893 gas per payout (as [expected](https://github.com/statechannels/statechannels/issues/3046)). If we get into the thousands of allocation items, we are now saving *millions* of gas overall.
